crates/ruff/src/commands/graph.rs

```diff
@@ -0,0 +1,159 @@
+use ruff_python_ast::{PySourceType, SourceType};
+use ruff_workspace::resolver::{python_files_in_path, ResolvedFile};
+use rustc_hash::FxHashMap;
+use std::collections::BTreeMap;
+use std::path::Path;
+use std::sync::Arc;
+
+/// Generate an import map.
+pub(crate) fn graph(args: GraphArgs, config_arguments: &ConfigArguments) -> Result<ExitStatus> {
+    // Construct the "default" settings. These are used when no `pyproject.toml`
+    // files are present, or files are injected from outside the hierarchy.
+    let pyproject_config = resolve(config_arguments, None)?;
+
+    // Find all Python files.
+    let files = resolve_default_files(args.files, false);
+    let (paths, resolver) = python_files_in_path(&files, &pyproject_config, config_arguments)?;
+
+    if paths.is_empty() {
+        warn_user_once!("No Python files found under the given path(s)");
+        return Ok(ExitStatus::Success);
+    }
+
+    // Resolve all package roots.
+    let package_roots = resolver
+        .package_roots(
+            &paths
+                .iter()
+                .flatten()
+                .map(ResolvedFile::path)
+                .collect::<Vec<_>>(),
+        )
+        .into_iter()
+        .map(|(path, package)| (path.to_path_buf(), package.map(Path::to_path_buf)))
+        .collect::<FxHashMap<_, _>>();
+
+    // Create a database for each source root.
+    let databases = package_roots
+        .values()
+        .filter_map(|package| package.as_deref())
+        .filter_map(|package| package.parent())
+        .map(Path::to_path_buf)
+        .map(|source_root| Ok((source_root.clone(), ModuleDb::from_src_root(source_root)?)))
+        .collect::<Result<BTreeMap<_, _>>>()?;
+
+    // Collect and resolve the imports for each file.
+    let result = Arc::new(std::sync::Mutex::new(Vec::new()));
+    let inner_result = Arc::clone(&result);
+
+    rayon::scope(move |scope| {
+        for resolved_file in &paths {
+            let Ok(resolved_file) = resolved_file else {
+                continue;
+            };
+
+            let path = resolved_file.path().to_path_buf();
+            let package = path
+                .parent()
+                .and_then(|parent| package_roots.get(parent))
+                .and_then(Clone::clone);
+            let Some(src_root) = package
+                .as_ref()
+                .and_then(|package| package.parent())
+                .map(Path::to_path_buf)
+            else {
+                debug!("Ignoring file outside of source root: {}", path.display());
+                continue;
+            };
+            let Some(db) = databases.get(&src_root).map(ModuleDb::snapshot) else {
+                continue;
+            };
+
+            // Resolve the per-file settings.
+            let settings = resolver.resolve(&path);
+            let string_imports = settings.graph.detect_string_imports;
+            let include_dependencies = settings.graph.include_dependencies.get(&path).cloned();
```

Comment:
It's a bit unfortunate that we need to clone this for every path.

Reply:
We're only cloning the return value though, which should be unique for every path.